scripts/Muon/GUI/Common/contexts/fitting_contexts/basic_fitting_context.py

```diff
@@ -0,0 +1,253 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+from Muon.GUI.Common.ADSHandler.ADS_calls import check_if_workspace_exist
+from Muon.GUI.Common.contexts.fitting_contexts.fitting_context import FittingContext
+
+
+class BasicFittingContext(FittingContext):
```

Comment:
This is a lot of boiler plate and can be replaced with a dataclass which will automatically create the getters and setters for you:

```python

from mantid.py36compat import dataclass


@dataclass
class AdjustmentStruct:
    # Be sure to include defaults if you don't want them to be mandatory to the constructor

    allow_double_pulse_fitting: bool = False
    _current_dataset_index: int = None  # This one is internal so we can have a property with validation
    dataset_names: List[str]  = [] # Include the type you expect in the list

    @property
    def current_dataset_index(self):
        return self._current_data_index

    @property.setter
    def current_dataset_index(self):
                if index is not None and index >= self.number_of_datasets:
            raise RuntimeError(f"The provided dataset index ({index}) is too large.")
        self._current_dataset_index = index
```
This allows us to have specific overrides by hiding attributes for things we want people to not get directly. Now all of the following calls will work with the above example

```
adj = AdjustmentStruct()
adj.allow_double_pulse_fitting  # returns False
adj.current_dataset_index = 1  # runs validators
adj.dataset_names = "a str"  # IDE would warn that were not using a list of str
```

Reply:
I am getting an error now that these are dataclasses:
```AttributeError: 'TFAsymmetryFittingContext' object has no attribute 'dataset_names'```

Will the TFAsymmetryFittingContext dataclass inherit the fields from the parent dataclasses?

Reply:
It was also complaining about there being no `default_factory` set so I am doing something like
```dataset_names: List[str] = field(default_factory=list, init=[])```

Is this right, it seems a bit complicated?

Reply:
Just thought of another question! Does it matter that the very base context (`FittingContext`) is not really a dataclass? At the moment the inheritance structure is `FittingContext`<--`BasicFittingContext`<--`GeneralFittingContext`<--`TFAsymmetryFittingContext`

Reply:
> I am getting an error now that these are dataclasses:
> `AttributeError: 'TFAsymmetryFittingContext' object has no attribute 'dataset_names'`


> It was also complaining about there being no `default_factory` set so I am doing something like
> `dataset_names: List[str] = field(default_factory=list, init=[])`
> 
> Is this right, it seems a bit complicated?
Yes, as this is a common bug (that even I've done). Because of the fact Python uses refs if you set default to `[]` every instance would use the same list, hence we have to use a factory.

```python
def myMethod(some_default=[None]):
    some_default.append(None)
    print(some_default)

myMethod()  # Prints `[None, None]`
myMethod()  # Prints `[None, None, None]`
```

This is something that factory tries to prevent basically.

> Just thought of another question! Does it matter that the very base context (`FittingContext`) is not really a dataclass? At the moment the inheritance structure is `FittingContext`<--`BasicFittingContext`<--`GeneralFittingContext`<--`TFAsymmetryFittingContext`

Do we need this inheritance structure if this class simply holds data? I'm not sure as I tend to refactor to make a data class purely hold data. That way you don't get any unexpected surprises where a property suddenly has side effects

Reply:
At the moment the `FittingContext` has not been refactored and so does more than just holds data - I would like to avoid refactoring this for now. 
If I were to separate each of these contexts from each other so they were no longer inheriting then there would be a lot of duplicate code as all of them would need the BasicFitting data fields and validation, and so on. I guess I could try combine the BasicFittingContext, GeneralFittingContext and TFAsymmetryFittingContext so it is one big dataclass which inherits from the `FittingContext`? However this would break away from the current inheritance structure of the view presenter and model files

Reply:
I think we would very quickly run the risk of a god class if we merged them into one. Each time a new fitting mode is added, we would need to extend the class. If we have the inheritance, we can have lots of smaller ones instead.